File: Lab5/VisualStudio/C++/matrix.cpp
#include "pch.h"
#include "matrix.h"

double matrix::operator()(int i, int j) { return mat[i][j]; }

matrix::matrix(const matrix& m) {
	int i, j;
	N = m.N;
	mat = new double* [N];
	for (i = 0; i < N; i++)
		mat[i] = new double[N];

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			mat[i][j] = m.mat[i][j];

}

matrix::matrix() {
	N = 0;
	mat = NULL;
}

matrix::~matrix() {
	int i;
	for (i = 0; i < N; i++) delete[] mat[i];
	if (mat != NULL) delete[] mat;
}

matrix::matrix(int ord) {
	int i, j;
	N = ord;
	mat = new double* [N];
	for (i = 0; i < N; i++)
		mat[i] = new double[N];

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			mat[i][j] = 0;
}

matrix::matrix(int k, int ord) {
	int i, j;
	N = ord;
	mat = new double* [N];
	for (i = 0; i < N; i++)
		mat[i] = new double[N];

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			mat[i][j] = i == j ? k : 0;
}

matrix::matrix(double* d, int ord) {
	int i, j;
	N = ord;
	mat = new double* [N];
	for (i = 0; i < N; i++)
		mat[i] = new double[N];

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			mat[i][j] = i == j ? d[i] : 0;

}


int matrix::order() {
	return N;
}

void matrix::set(int i, int j, double val) {
	if (i < 1) throw Exept("wrong first parameter");
	if (j < 1) throw Exept("wrong second parameter");
	mat[i - 1][j - 1] = val;
}

matrix& matrix::operator =(const matrix& m) {
	int i, j;

	for (i = 0; i < N; i++) delete[] mat[i];
	if (mat != NULL) delete[] mat;
	N = m.N;
	mat = new double* [N];
	for (i = 0; i < N; i++)
		mat[i] = new double[N];

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			mat[i][j] = m.mat[i][j];

	return *this;

}

std::ostream& operator << (std::ostream& out, matrix m) {
	int i, j;

	for (i = 0; i < m.order(); ++i) {
		for (j = 0; j < m.order(); ++j)
			out << m(i, j) << ' ';
		out << std::endl;
	}

	return out;
}

//сложение матриц
matrix matrix::operator +(matrix& m) {
	int i, j;

	if (N != m.N) throw Exept("matrix 1 and 2 have different size");

	matrix out(N);

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			out.mat[i][j] = mat[i][j] + m.mat[i][j];

	return out;
}
matrix& matrix::operator +=(matrix& m) {
	*this = *this + m;

	return *this;
}

//вычитание матриц
matrix matrix::operator -(matrix& m) {
	int i, j;

	if (N != m.N) throw Exept("matrix 1 and 2 have different size");

	matrix out(N);

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			out.mat[i][j] = mat[i][j] - m.mat[i][j];

	return out;
}
matrix& matrix::operator -=(matrix& m) {
	*this = *this - m;

	return *this;
}

//умножение матриц
matrix matrix::operator *(matrix& m) {
	int i, j, k;
	if (N != m.N) throw Exept("can't multiply matrix 1 and 2");
	matrix out(N);

	for (i = 0; i < N; i++)
		for (j = 0; j < N; j++)
			for (k = 0; k < N; k++)
				out.mat[i][j] += mat[i][k] * m.mat[k][j];

	return out;
}
matrix& matrix::operator*=(matrix& m) {
	*this = *this * m;

	return *this;
}

//умножение матрицы на вектор
vectorb matrix::operator *(vectorb& v) {
	if (N != v.blockOrder() || v.order() > 1) throw Exept("can't multiply matrix to vector");
	vectorb out(1, v.blockOrder());

	for (int i = 0; i < out.blockOrder(); i++) {
		for (int j = 0; j < out.blockOrder(); j++)
			out[0][i] += mat[i][j] * v[0][j];
	}

	return out;
}
vectorb matrix::operator *(double* v) {
	vectorb out(1, N);

	for (int i = 0; i < out.blockOrder(); i++) {
		for (int j = 0; j < out.blockOrder(); j++)
			out[0][i] += mat[i][j] * v[j];
	}

	return out;
}

double* matrix::operator [] (int i) {
	return mat[i];
};

File: Lab5/VisualStudio/Csharp/matrix.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CsharpTest
{
    public class matrix
    {
        private int N;
        public int GetN { get => this.N; }

        private double[,] mat;

        public void Function(Action<int, int> f)
        {
            for (var i = 0; i < N; i++)
                for (var j = 0; j < N; j++)
                    f(i, j);
        }
        public matrix()
        {
            N = 0;
            mat = null;
        }

        public matrix(int ord)
        {
            N = ord;
            mat = new double[N, N];
            this.Function((i, j) => mat[i, j] = 0);
        }

        public matrix(int k, int ord)
        {
            N = ord;
            mat = new double[N, N];
            this.Function((i, j) => mat[i, j] = i == j ? k : 0);
        }
        public matrix(double[] d, int ord)
        {
            N = ord;
            mat = new double[N, N];
            this.Function((i, j) => mat[i, j] = i == j ? d[i] : 0);
        }

        public double this[int x, int y]
        {
            get
            {
                return this.mat[x, y];
            }
            set
            {
                this.mat[x, y] = value;
            }
        }


        public static matrix operator +(matrix one, matrix two)
        {
            if (one.N != two.N)
                throw new ArgumentException("matrixes have different dimensions");
            var res = new matrix(one.N);
            res.Function((i, j) => res[i, j] = one[i, j] + two[i, j]);
            return res;
        }

        public static matrix operator -(matrix one, matrix two)
        {
            if (one.N != two.N)
                throw new ArgumentException("matrixes have different dimensions");
            var res = new matrix(one.N);
            res.Function((i, j) => res[i, j] = one[i, j] - two[i, j]);
            return res;
        }

        public static matrix operator *(matrix one, matrix two)
        {
            if (one.N != two.N)
                throw new ArgumentException("matrixes have different dimensions");
            var res = new matrix(one.N);
            res.Function((i, j) =>
            {
                for (var k = 0; k < one.N; k++)
                    res[i, j] += one[i, k] * two[k, j];
            });
            return res;
        }

        public static vectorb operator *(matrix one, vectorb v)
        {
            if (one.N != v.GetBlockN || v.GetN > 1)
                throw new ArgumentException("different dimensions");
            vectorb res = new vectorb(1, v.GetBlockN);
            for (var i = 0; i < v.GetBlockN; i++)
                for (var j = 0; j < v.GetBlockN; j++)
                    res[0][j] += one[i, j] * v[0][j];

            return res;
        }
        public static vectorb operator *(matrix one, double[] d)
        {
            vectorb res = new vectorb(1, one.N);
            for (var i = 0; i < one.N; i++)
                for (var j = 0; j < one.N; j++)
                    res[0][j] += one[i, j] * d[j];

            return res;
        }
        public override string ToString()
        {
            string s = "";
            for (var i = 0; i < N; i++)
            {
                for (var j = 0; j < N; j++)
                    s += mat[i, j] + " ";
                s += "\n";
            }
            return s;
        }

        public static matrix inverse(matrix m)
        {
            matrix M = new matrix(m.GetN);
            for (int i = 0; i < m.GetN; i++)
                M[i, i] = 1 / m[i, i];
            return M;
        }

    }
}
